magefiles/run.go

```diff
@@ -0,0 +1,209 @@
+//go:build mage
+// +build mage
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strings"
+
+	"github.com/magefile/mage/mg"
+	"github.com/magefile/mage/sh"
+)
+
+// Create and run Kubernetes cluster with two nodes, Linux and Windows.
+func Run() error {
+	mg.SerialDeps(startup, checkVagrant, checkClusterNotRunning)
+
+	var err error
+
+	log.Println("Creating .lock directory")
+	_, err = os.Stat(".lock")
+	if !os.IsNotExist(err) {
+		err = os.RemoveAll(".lock")
+		if err != nil {
+			return err
+		}
+	}
+	err = os.Mkdir(".lock", os.ModePerm)
+	if err != nil {
+		return err
+	}
+
+	// Linux node //////////////////////////////////////////////////////////////
+	log.Println("Creating control plane Linux node")
+	err = sh.Run("vagrant", "up", "controlplane")
+	if err != nil {
+		return err
+	}
+
+	mustSetVagrantPrivateKeyPermissions()
+
+	err = sh.Run("vagrant", "status")
+	if err != nil {
+		return err
+	}
+
+	// Windows node ////////////////////////////////////////////////////////////
+	log.Println("Creating worker Windows node")
+	log.Println("##########################################################")
+	log.Println("Retry vagrant up if the first time the Windows node failed")
+	log.Println("##########################################################")
+
+	const maxAttempts = 10 // TODO: How many to allow? Make configurable via variables.yaml?
+	for i := 1; i < maxAttempts; i++ {
+		log.Printf("vagrant status winw1 - attempt %d of %d", i, maxAttempts)
+		output, err := sh.Output("vagrant", "status", "winw1")
+		if err != nil {
+			return err
+		}
+		status := extractMachineStatus(output, "winw1")
+		log.Println("winw1", status)
+		if strings.Contains(status, "running") {
+			break
+		}
+		log.Printf("vagrant up winw1 - attempt %d of %d", i, maxAttempts)
+		err = sh.Run("vagrant", "up", "winw1")
+		if err != nil {
+			return err
+		}
```

Comment:
we should add some sleeps in between tries, so we don't fail immediately.

Reply:
Possibly, on the other hand there is this up & status & provision loop (which comes from the original `Makefile`).
After some testing I think the follow-up with `vagrant provision` may not be necessary at all and then some sleep may help indeed.

For now, the magefiles reproduce sequence from `Makefile`.
I only added number of max allowed attempts, which is configurable now in `settings.[local.]yaml`.